File: go/2022/day1.go
package _2022

import (
	"advent/utils"
	"fmt"
	"sort"
	"strings"
)

func getCalories(raw string) []int {
	elves := strings.Split(raw, "\n\n")
	calories := make([]int, len(elves))
	for ii, elf := range elves {
		elfCalories := 0
		for _, s := range strings.Split(elf, "\n") {
			elfCalories += utils.ParseInt(s)
		}
		calories[ii] = elfCalories
	}
	return calories
}

func Day1Part1(input string) string {
	maxCalories := utils.Max(getCalories(input))
	return fmt.Sprint(maxCalories)
}

func Day1Part2(input string) string {
	calories := getCalories(input)
	sort.Slice(
		calories,
		func(a, b int) bool { return calories[a] > calories[b] },
	)
	sum := utils.Sum(calories[:3])
	return fmt.Sprint(sum)
}

File: python/puzzles/y2022/day1.py
import utils


def get_calories(raw: str):
    elves = raw.split("\n\n")
    return [sum(map(int, elf.split("\n"))) for elf in elves]


@utils.profile
def part1(raw: str):
    return max(get_calories(raw))


@utils.profile
def part2(raw: str):
    return sum(sorted(get_calories(raw), reverse=True)[:3])


if __name__ == "__main__":
    raw = utils.load_puzzle_input("2022/day1")
    assert part1(raw) == 66186
    assert part2(raw) == 196804
